onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -51,8 +54,28 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       node->SetExecutionProviderType(kCpuExecutionProvider);
     }
 
+    // Check whether the CPU EP has a kernel for this node. Stop proceeding if it doesn't.
+    std::unique_ptr<CPUExecutionProvider> cpu_execution_provider =
+        onnxruntime::make_unique<CPUExecutionProvider>(CPUExecutionProviderInfo());
+
+    auto cpu_kernel_found_for_node =
+        KernelRegistry::HasImplementationOf(*cpu_execution_provider->GetKernelRegistry(), *node, kCpuExecutionProvider);
+
+    if (!cpu_kernel_found_for_node) {
+      LOGS(logger, WARNING) << "Could not find a CPU kernel and hence "
+                            << "can't constant fold " << node->OpType() << " node '" << node->Name() << "'";
+
+      // Revert the EP change. We won't be proceeding with constant folding for this node.
+      if (!cpu_ep) {
+        node->SetExecutionProviderType(ep_type);
+      }
+
+      // Move on to the next candidate node
```

Comment:
Can the check for a kernel move above where we first override the node's EP so we don't need to undo that here?

Reply:
When I checked the logic of HasImplementationOf(), it seemed like the node's EP (if set) is what will be matched with the kernel defs held by the kernel registry, so we have to mutate it to CPU EP or unset it and depend on the passed in provider_type. Either approach requires reverting a mutation.